### ISSUE TITLE
Lecture 3: Liskov's Substitutability Principle

### ISSUE THREAD

**Issue by shermanyuen** [nus-cs2030/2021-s1#32]
Hi, I have a question about Lecture 3 materials, in particular on Liskov's Substitutability Principle.

**I would like to ask what is the intended outcome of the method scaleBy?**

1. If `UnitCircle uc = new UnitCircle(p)`, is `uc.scaleBy(2)` supposed to return back a `UnitCircle` with the same properties (no change in radius), return a `Circle` (but not `UnitCircle`) with radius 1, or return a `Circle` with radius 2?
- In Slide 6, in the comment of the 3rd jshell `// uc remains :)`, it suggests that returning a `UnitCircle` is correct.
- In Slide 7, in the comment of the 2nd jshell `// Still valid?`, it suggests that returning a `UnitCircle` is wrong, by the virtue of the `?`.
- In the lecture recording from 53:20 to 54:35, the lecturer says that we should return back a `UnitCircle` if the factor is 1, and return back a `Circle` with the radius changed if the factor is not 1. So he is suggesting that returning a `Circle` with radius 2 is correct.

2. If `Circle c = new UnitCircle(p)`, is `c.scaleBy(2)` supposed to return back a `UnitCircle` with the same properties (no change in radius), return a `Circle` (but not `UnitCircle`) with radius 1, or return a `Circle` with radius 2?
It is a similar question to Q1, but I am wondering what is the intended outcome. Only after understanding what the intended outcome is then we can consider how the code should be written.

**Comment by darrenhoon** [comment]
1. I believe in the lecture Prof Henry proceeded without using Circle since it requires a Point too. But I believe the spirit is such that it returns a Circle with radius 2 (since a UnitCircle implies that it has radius 1). otherwise calling it UnitCircle would be confusing to both the implementer and the client (imagine if you are the client, calling what is supposed to be a **Unit**Circle but instead getting back a circle with radius 2)

2. In this case, the Circle c object "looks" at the entire UnitCircle from the parent class's perspective (ie Circle). But because the scaleBy in the UnitCircle is overriding the parent's scaleBy, when the c object calls the parent's scaleBy method, it is actually the child's scaleBy method. thus it returns a UnitCircle of radius 1.

**Comment by shermanyuen** [comment]
Hi Darren, thanks for your reply.
Actually I am even more confused by your response. Let me try to summarise what you are saying.

Are you saying that:
1. If `UnitCircle uc = new UnitCircle(p)`, then `uc.scaleBy(2)` is supposed to return a `Circle` with radius 2?
2. If `Circle c = new UnitCircle(p)`, then `c.scaleBy(2)` is supposed to return a `UnitCircle` with radius 1?

If that is what you are saying, then I am even more confused by what we want `scaleBy` to do...

**Comment by darrenhoon** [comment]
The idea of LSP, is that you can do (Circle) UnitCircle uc = new UnitCircle( new Point(0,0) ) and retain all the functions of UnitCircle (ie if you do scaleBy(5), then the uc will actually scale by 5)


1. I believe having a scaleBy function actually violates LSP as LSP says that if you type-casted UnitCircle uc with Circle, ie (Circle) uc, then uc should behave like a Circle instead of a UnitCircle. But it is not possible for you return a scaled Circle due to the scaleBy method in UnitCircle overriding the Circle's scaleBy method. I believe the code was:


```
class UnitCircle extends Circle{
    UnitCircle(Point centre){
        super(centre, 1.0);
    }

    UnitCircle scaleBy(double factor){
        return this;
    }
}
```

Now, when you think about it, if you removed the UnitCircle's scaleBy method, the method uc.scaleBy calls the parent class's scaleBy method, which is the following:

```
Circle scaleBy(double factor){
        return new Circle(this.centre, this.radius*factor);
    }
```

But this also means that if you do (UnitCircle) uc.scaleBy(5), what it returns is not a UnitCircle but a Circle with the scaled parameters, even without the typecasting of Circle to uc.

This is why the code violates LSP because there is no way to edit the code such that you can allow for the UnitCircle to have a scaleBy method that returns itself but when the Parent class is type-casted over it, it can change its scale (at least from what we learn so far).

Apologies if the previous comment was confusing as I was taking it from the perspective of the client expecting scaleBy to work, irregardless of the class being either UnitCircle or Circle (which in this instance, the solution would be to remove the scaleBy method in UnitCircle).



2. So this part builds on the above, because if you don't do anything to the UnitCircle's scaleBy method (ie just leave it there), then Circle c = new UnitCircle(new Point(0,0)) is to look at the UnitCircle from the point of view of a Circle. To use Prof Terence's example from Recitation 2, below is how it will look like for this case:

![Screenshot_1](https://user-images.githubusercontent.com/34526101/92298360-81be4380-ef7a-11ea-955f-5150d118be33.jpg)

so back to your original Question 2, yes it will return a UnitCircle object, but with the view of a Circle. Below are more screenshots to depict the behaviour:

![Screenshot_2](https://user-images.githubusercontent.com/34526101/92298370-91d62300-ef7a-11ea-8e8c-2f0e5fe48959.jpg)
_shows that c will return a UnitCircle object with a view from the Circle class_

![Screenshot_3](https://user-images.githubusercontent.com/34526101/92298376-98649a80-ef7a-11ea-9254-02816660b61e.jpg)
_to further illustrate the whole concept of viewing the UnitCircle from the lens of the Circle_


Feel free to ask more since we're all learning together ^^

**Comment by shermanyuen** [comment]
Hi Darren, again thank you for your answer to my question. I appreciate your effort but I am actually a little lost in the whole discussion. 

My question is not really about what is LSP and not about whether certain codes violate LSP or not. My question is also not about what certain codes do to the objects. (Maybe these could be my questions later, but for now I am trying to understand something else.)

My question is really, what do we, as the programmer, want `scaleBy` to do? Perhaps it is easier for me to list down some options and then you can pick accordingly. Let me try to be more specific.

1. For `UnitCircle uc = new UnitCircle(p)` where `uc` has a compile-time type of `UnitCircle` and a run-time type of `UnitCircle`, should `uc.scaleBy(2)` return:
1A. a compile-time type and run-time type `UnitCircle` of radius 1? or
1B. a compile-time type `Circle` and a run-time type `UnitCircle` of radius 1? or
1C. a compile-time type and run-time type `Circle` of radius 2? or
1D. none of the above? (if so, what is it?)

2. For `Circle c = new UnitCircle(p)` where `c` has a compile-time type of `Circle` and a run-time type of `UnitCircle`, should `c.scaleBy(2)` return:
2A. a compile-time type and run-time type `UnitCircle` of radius 1? or
2B. a compile-time type `Circle` and a run-time type `UnitCircle` of radius 1? or
2C. a compile-time type and run-time type `Circle` of radius 2? or
2D. none of the above? (if so, what is it?)

The main reason why I am asking this question is because I found contradictions in the lecture notes and the lecture recording, as described in my original post.

My original understanding is (1A & 2B), but then later my understanding is (1C & 2C). Now I am not sure which one is correct.

---

My follow-up question (later) will probably be something like: will any code satisfying the answer to my original question, say (1X & 2Y), violate the LSP or not?

**Comment by HolmesJJ** [comment]
I try to summarize the Liskov's Substitutability Principle with a simple explanation:
**Subclass can extend the functions of superclass, but do not change the original functions of the superclass.**

1. The method of the subclasses can **override** the **abstract method** of the superclass, but **try not to** override the non-abstract methods of the superclass. A more accurate explanation is: for non-abstract method, the specific behavior of the method can be changed, but the purpose of the method **should not** be changed. (For example, the purpose of the scaleBy() method in UnitCircle has been changed, which violates LSP.)
2. Subclasses can add its own unique methods. (That's because the subclass is more specialized than the superclass.)
3. When the method of the subclass **overloading** the method of the superclass, the **pre-condition** of the method (the parameters) of the subclass **need to be more relaxed** than the superclass. (That's because **overloading** will call the more specialized method.)
4. When the method of the subclass **override** the **abstract method** of the superclass, the **post-condition** of the method (return value) of the subclass **need to be stricter** than the superclass. (Otherwise, the program cannot be compiled, because the **downcasting** of the class is **not safe**. Note that this refers to **override** the **abstract method**, non-abstract methods try not to be **override**. )

So, go back to the UnitCircle and Circle example, there have different characteristics between UnitCircle (cannot be scaled) and Circle (can be scaled), so UnitCircle **should not** inherit Circle, they can each inherit a lower base class Shape.

We need to know that while inheritance brings great convenience to programming, it also brings drawbacks. The Liskov's Substitutability Principle only tells us that we should pay attention to the above problems when designing a program, but it does not stipulate what we must do. Even if the LSP is violated, the program still can run well. 

So what are the consequences if we don't follow the LSP?
**The chances of bugs in your code will greatly increase.**

**Comment by waynetee** [comment]
@shermanyuen I think this is a pretty well thought out question, let me give my opinion.

You asked what should scaleBy do?

What we expect code to do is subjective, but I think it's fair to expect code to follow principles such as LSP.
If UnitCircle extends Circle, then by LSP, as Darren and JiaJun explained, you can eliminate a couple of options from your list.

By LSP, we know that the return type of the UnitCircle's scaleBy method, whether you override it not, has to be of type Circle.

I'll like to note that it is possible for the UnitCircle's scaleBy method to return a UnitCircle if the factor is 1. However, I think the benefits are minimal since the return type in the method header has to remain as a circle.

Notice that LSP can tell us that something is wrong but not what is right or what is best.

So what is the intended behaviour of scaleBy?

If you give me a problem e.g. "Model a world with scalable shapes and fixed shapes...."  I can propose a solution.
If you give me a solution to a problem, I can guess what the problem is.
But if you give me neither and ask me what is the problem, I can only tell you what cannot be a good solution given principles such as LSP.

If you want to model a world where UnitCircles can scale to become Circles, then you can maintain the inheritance and not override scaleBy.
If you want to model a world with scalable circles and fixed circles, you can consider `ScalableCircle extends Circle`.
JiaJun proposed a good alternative implementation as well.

To summarise my response, there is not enough code/information to infer exactly what the problem is, and hence, as you mentioned, without knowing the problem we cannot derive the solution. But we can still try out different scenarios and consider how we might solve them.

Hope this helps, do share your thoughts.

**Comment by darrenhoon** [comment]
Hi @shermanyuen, apologies for the delayed reply. I believe your confusion arises from slides 6 and 7, where it seems to be providing contradictory information. However, I believe what Prof Henry was trying to do was to get the audience to ponder and evaluate which implementation is better, and your question in this thread has greatly enhanced the discussion, as seen in @waynetee and @HolmesJJ 's replies which elaborate upon the crux of the issue - which is what we as programmers actually what the implementation to be.

Both implementations have their draw-backs, with either losing clarity (eg UnitCircle? Circle? ScalableCircle? ColourfulCircle?) or losing functionality (eg UnitCircle can never be scaled, which begs the question of why do we even let UnitCircle to have the scaleBy function to begin with or how it fits into the overall system of the client-implementer relationship). The above discussion by us attempts to reason, using LSP, which is the more preferred method. But as you know, a lot of things are variable upon the specific context in which it is applied (eg if you work with a client who explicitly wants his classes to be "monsters" as Prof Terence quips in his recitation for the sake of convenience, it probably will violate LSP and even the SOLID principle in general)

To have a direct answer to your question, and this is my own opinion (thus do not treat it as the absolute answer), below are my choices and reasoning behind it:
1. 1A.
To call it a UnitCircle when it can behave in a scalable manner would potentially lead to more confusion in the future. Try and think of it from the perspective of a client, where I tell you we have UnitCircle, Circle, ScalableCircle. You would probably ask what is the difference between them, and if there is no discernable difference between all 3, a logically reply would be then why bother differentiating all 3? At least this is my take, since a Circle can always have just a unit radius, where the burden of interpretation of whether an object is a unit object lies incumbent on the user. If there exists a UnitCircle (which in this case it does), it should have a difference from a normal Circle or else I'd rather just have the Circle object (since UnitCircle instantiates differently from a Circle so I would have to do more work later to create for-loops that accomodate to the different instantiation)

2. 2C.
This is a tricky one. I'm going to assume the context is that this time, I am the client and I typecasted Circle to an array of Circle objects, where this example is just one instance of the object actually being a UnitCircle rather than a Circle (but idk because I am looking at it without the knowledge of there being a difference between the classes in the array). What this means is that I would expect UnitCircle to behave like a Circle (ie scaleBy actually scales the circle), but due to the implementation in 1, what I will get back is a UnitCircle, which does not meet my expectations.

This goes to show that it is probably a bad idea to go about implementing Circles/UnitCircles in such a way that UnitCircle is fixed if the client expects it to be behave like a Circle. On the other hand, if you made it such that UnitCircle essentially behaves like a Circle, then what is the point of creating a subclass (if it looks like a duck, swims like a duck, and quacks like a duck, then it probably is a duck).



Perhaps any of the profs or TAs can weigh in on this discussion :))

**Comment by charoi** [comment]
Hi @shermanyuen ! Seen a lot of helpful answers about the LSP actually. I was also personally really confused by LSP, so here's a video I watched on Youtube trying to simplify LSP for everyone!
[LSP Video](https://youtu.be/Mmy1EUKC_iE)

Assuming you understand LSP, I would say the best way to look at it is from the perspective of UnitCircle's purpose, which is that it is a circle of radius one. Based on my understanding, any method created, be it in run/compile-time of Circle or UnitCircle, should not be altering the radius, else it will violate the LSP rule. Other methods such as volume, area, or perimeter can be overridden by the very basis of a unit circle being of a fixed length makes it violable to LSP whenever a method concerning altering radius, such as scaleBy, is defined.

Also, my personal answer would be 1A & 2B, and as a programmer, I think it doesn't make sense to have a **scaleBy** method and a **unitCircle** subclass due to the oft-mentioned LSP in this discussion, which is why everyone is bringing it up.

Hope this helps!
